File: 03-provider-registry/nodejs/src/providers/openai.ts
import OpenAI from 'openai';
import type { Message, Provider } from './types.ts';

export function createOpenAIProvider(): Provider {
  const client = new OpenAI({
    apiKey: process.env['OPENAI_API_KEY'],
    baseURL: process.env['OPENAI_API_BASE_URL'],
  });
  const model = process.env['OPENAI_MODEL'] ?? 'gpt-4o';

  return {
    name: 'openai',
    contextWindow: 128_000,

    async chat(messages: Message[]): Promise<string> {
      const completion = await client.chat.completions.create({
        model,
        messages: messages.map(m => ({ role: m.role, content: m.content })),
      });
      return completion.choices[0].message.content ?? '';
    },
  };
}

File: 03-provider-registry/golang/providers/openai.go
package providers

import (
	"context"
	"os"

	openai "github.com/sashabaranov/go-openai"
)

type openAIProvider struct {
	client *openai.Client
	model  string
}

func NewOpenAIProvider() Provider {
	config := openai.DefaultConfig(os.Getenv("OPENAI_API_KEY"))
	if base := os.Getenv("OPENAI_API_BASE_URL"); base != "" {
		config.BaseURL = base
	}
	model := os.Getenv("OPENAI_MODEL")
	if model == "" {
		model = "gpt-4o"
	}
	return &openAIProvider{client: openai.NewClientWithConfig(config), model: model}
}

func (p *openAIProvider) Name() string         { return "openai" }
func (p *openAIProvider) ContextWindow() int   { return 128_000 }

func (p *openAIProvider) Chat(ctx context.Context, messages []Message) (string, error) {
	// OpenAI format is a direct mapping from the unified Message type.
	msgs := make([]openai.ChatCompletionMessage, len(messages))
	for i, m := range messages {
		msgs[i] = openai.ChatCompletionMessage{Role: m.Role, Content: m.Content}
	}

	resp, err := p.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:    p.model,
		Messages: msgs,
	})
	if err != nil {
		return "", err
	}
	return resp.Choices[0].Message.Content, nil
}
